samples/example_vptr/example_vptr.cpp

```diff
@@ -0,0 +1,129 @@
+   * associated with those objects is completed. */
+  {
+    queue myQueue;
+    cl::sycl::codeplay::PointerMapper pMap;
+
+    /* Create device-only 2D buffers of floats for the matrices. */
+    /*
+    buffer<float, 2> a(range<2>{N, M});
+    buffer<float, 2> b(range<2>{N, M});
+    buffer<float, 2> c(range<2>{N, M});
+    */
+    float* a = static_cast<float*>(SYCLmalloc(N * M * sizeof(float), pMap));
+    float* b = static_cast<float*>(SYCLmalloc(N * M * sizeof(float), pMap));
+    float* c = static_cast<float*>(SYCLmalloc(N * M * sizeof(float), pMap));
+
+    /* This kernel enqueue will initialise buffer a. The accessor "A" has
+     * write access. */
+    myQueue.submit([&](handler& cgh) {
+      auto A = pMap.get_access<access::mode::write,
+                               access::target::global_buffer, float>(a, cgh);
+      cgh.parallel_for<class init_a>(
+          range<1>(N * M), [=](id<1> index) { A[index] = index[0] * 2; });
+    });
+
+    /* This kernel enqueue will likewise initialise buffer b. The only
+     * accessor it specifies is a write accessor to b, so the runtime
+     * can use this information to recognise that these kernels are
+     * actually independent of each other. Therefore, they can be enqueued
+     * to the device with no dependencies between each other. */
+    myQueue.submit([&](handler& cgh) {
+      auto B = pMap.get_access<access::mode::write,
+                               access::target::global_buffer, float>(b, cgh);
+      cgh.parallel_for<class init_b>(
+          range<1>{N * M}, [=](id<1> index) { B[index] = index[0] * 2014; });
+    });
+
+    /* This kernel will actually perform the computation C = A * B. Since
+     * A and B are only read from, we specify read accessors for those two
+     * buffers, which the SYCL runtime recognises as a dependency on the
+     * previous kernels. If the data were initialised on a different device,
+     * or on the host, the SYCL runtime would ensure that the data were
+     * copied between contexts etc. properly. */
+    myQueue.submit([&](handler& cgh) {
+      auto A = pMap.get_access<access::mode::write,
+                               access::target::global_buffer, float>(a, cgh);
+      auto B = pMap.get_access<access::mode::write,
+                               access::target::global_buffer, float>(b, cgh);
+      auto C = pMap.get_access<access::mode::write,
+                               access::target::global_buffer, float>(c, cgh);
+      cgh.parallel_for<class matrix_add>(range<1>{N * M}, [=](id<1> index) {
+        C[index] = A[index] + B[index];
+      });
+    });
+
+    /* A host accessor will copy data from the device and, under most
+     * circumstances, allocate space for it for the user (it will not
+     * allocate space when the map_allocator is used and an initial host
+     * pointer is provided, as this instructs the runtime to map the data
+     * into the host's memory). Since this code is attempting to access
```

Comment:
Is there a sample using a map allocator? Should that be linked here?
It feels a bit strange to me that you are bringing up something that isn't used here and doesn't really have any direct relevance to this sample.

Reply:
Removed irrelevant comments. 0c2d66be8f9dee86bcaa729af98eca7ff31cbfa0